### PR TITLE
add Red Hat converter

### PR DESCRIPTION
**Pull request by jasinner**
Add a Red Hat CSAF to OSV converter tool used by Red Hat to produce the Red Hat OSV data to be hosted at https://security.access.redhat.com/data/osv

This is slightly modified version of the sample code originally review at https://github.com/andrewpollock/rhcsaf2osv

### CODE REVIEW

File: tools/redhat/csaf.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+import json
+from typing import Any, Iterable
+
+class Remediation:
```
Comment (andrewpollock) [inline tools/redhat/csaf.py:4]: FYI I was schooling myself on https://pylint.readthedocs.io/en/stable/user_guide/messages/refactor/too-few-public-methods.html and https://docs.python.org/3/library/dataclasses.html and asked Gemini to rewrite this as a dataclass and this is what it gave me:

```
from dataclasses import dataclass, field
from typing import Dict

@dataclass
class Remediation:
    """
    class to handle remediation advice in CSAF data
    """

    csaf_product_id: str
    cpes: Dict[str, str]
    purls: Dict[str, str]
    product: str = field(init=False)
    product_version: str = field(init=False)
    component: str = field(init=False)
    fixed_version: str = field(init=False)
    purl: str = field(init=False)
    cpe: str = field(init=False)

    def __post_init__(self):
        if ":" not in self.csaf_product_id:
            raise ValueError(f"Did not find ':' in product_id: {self.csaf_product_id}")
        (self.product, self.product_version) = self.csaf_product_id.split(":", maxsplit=1)

        # ... (rest of the logic from the original __init__ method, adjusted as needed)

        if not self.purl.startswith("pkg:rpm/"):
            raise ValueError("Non RPM remediations are not supported in OSV at this time")
``

Looks a bit nicer?

File: tools/redhat/csaf.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            raise ValueError(f"Did not find {csaf_product_id} in product branches")
+
+        # There are many pkg:oci/ remediations in Red Hat data. However there are no strict
+        # rules enforced on versioning Red Hat containers, therefore we cant compare container
+        # versions to each other with 100% accuracy at this time.
+        if not self.purl.startswith("pkg:rpm/"):
+            raise ValueError("Non RPM remediations are not supported in OSV at this time")
+
+
+class Vulnerability:
```
Comment (andrewpollock) [inline tools/redhat/csaf.py:46]: Gemini's dataclass alternative:

```
from dataclasses import dataclass, field
from typing import Dict, Any, List

@dataclass
class Vulnerability:
    """
    class to handle vulnerability information
    """

    csaf_vuln: Dict[str, Any]
    cpes: Dict[str, str]
    purls: Dict[str, str]
    cve_id: str = field(init=False)
    cvss_v3_vector: str = field(init=False, default=None) 
    cvss_v3_base_score: float = field(init=False, default=None)
    references: List[Dict[str, str]] = field(init=False)
    remediations: List[Remediation] = field(init=False, default_factory=list)

    def __post_init__(self):
        self.cve_id = self.csaf_vuln["cve"]
        for score in self.csaf_vuln.get("scores", []):
            if "cvss_v3" in score:
                self.cvss_v3_vector = score["cvss_v3"]["vectorString"]
                self.cvss_v3_base_score = score["cvss_v3"]["baseScore"]
        self.references = self.csaf_vuln["references"]

        for product_id in self.csaf_vuln["product_status"]["fixed"]:
            try:
                self.remediations.append(Remediation(product_id, self.cpes, self.purls))
            except ValueError as e:
                print(f"Could not parse product_id: {product_id}. {e}")
```

File: tools/redhat/osv.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+class OSVEncoder(JSONEncoder):
+    """Encodes OSV objects into JSON format"""
+
+    def default(self, o):
+        if isinstance(o, Event):
+            return o.encode_json()
+        return o.__dict__
+
+
+class Event:
```
Comment (andrewpollock) [inline tools/redhat/osv.py:16]: Another opportunity for a dataclass:

```
from dataclasses import dataclass
from typing import Literal

@dataclass
class Event:
    """
    Class to hold event information for a Range. 
    Advisories for Red Hat RPM based products always assume all previous versions are affected.
    """

    INTRODUCED: Literal["introduced"] = "introduced"
    FIXED: Literal["fixed"] = "fixed"

    event_type: Literal["introduced", "fixed"]
    version: str = "0"

    def __post_init__(self):
        if self.event_type not in (self.INTRODUCED, self.FIXED):
            raise ValueError(f"Expected one of {(self.INTRODUCED, self.FIXED)} for type. Got {self.event_type}")

    def encode_json(self):
        """
        Custom JSON encoding for event type which changes attribute name 
        depending on the type of event e.g., introduced or fixed
        """
        if self.event_type == self.INTRODUCED:
            return {self.INTRODUCED: self.version}
        if self.event_type == self.FIXED:
            return {self.FIXED: self.version}
        raise ValueError("Unexpected event_type for Event")
```

File: tools/redhat/osv.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        event eg. introduced or fixed
+        """
+        if self.event_type == Event.INTRODUCED:
+            return {Event.INTRODUCED: self.version}
+        if self.event_type == Event.FIXED:
+            return {Event.FIXED: self.version}
+        raise ValueError("Unexpected event_type for Event")
+
+
+class Range:
```
Comment (andrewpollock) [inline tools/redhat/osv.py:48]: Another dataclass opportunity

```
from dataclasses import dataclass, field
from typing import List

@dataclass
class Range:
    """
    Class to hold range information for a Package. 
    Ecosystem here refers to RPM versions as defined in 
    https://github.com/rpm-software-management/rpm/blob/master/rpmio/rpmvercmp.c
    """

    fixed: str
    type: str = field(init=False, default="ECOSYSTEM")
    events: List[Event] = field(init=False) 

    def __post_init__(self):
        self.events = [Event("introduced"), Event("fixed", self.fixed)]
```

File: tools/redhat/osv.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    # pylint: disable=too-few-public-methods
+    # This class encapsulates version range types as 'ECOSYSTEM' type
+    # Only initialization is required because data retrieval is via JSON encoding
+
+    def __init__(self, fixed: str):
+        self.type = "ECOSYSTEM"
+        self.events = [Event("introduced"), Event("fixed", fixed)]
+
+
+class Package:
```
Comment (andrewpollock) [inline tools/redhat/osv.py:63]: Another :-)

```
import re
from dataclasses import dataclass, field

@dataclass
class Package:
    """
    Class to hold package data for an Affect. 
    Expects an ecosystem string that starts with CPE_PATTERN.
    Replaces the CPE prefix 'redhat' part with 'Red Hat' 
    to match more closely with other ecosystem identifiers in the OSV database
    """

    CPE_PATTERN: re.Pattern = field(init=False, default=re.compile(r"cpe:/[oa]:(redhat)"))
    name: str
    ecosystem: str
    purl: str

    def __post_init__(self):
        if not self.CPE_PATTERN.match(self.ecosystem):
            raise ValueError(f"Got unsupported ecosystem: {self.ecosystem}")
        self.ecosystem = f"Red Hat{self.CPE_PATTERN.split(self.ecosystem, maxsplit=1)[-1]}"
```

File: tools/redhat/osv.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+    def __init__(self, name: str, ecosystem: str, purl: str):
+        self.name = name
+        if not self.CPE_PATTERN.match(ecosystem):
+            raise ValueError(f"Got unsupported ecosystem: {ecosystem}")
+        self.ecosystem = f"Red Hat{self.CPE_PATTERN.split(ecosystem, maxsplit=1)[-1]}"
+        self.purl = purl
+
+
+class Affected:
```
Comment (andrewpollock) [inline tools/redhat/osv.py:85]: Another dataclass

```
from dataclasses import dataclass, field
from typing import List

@dataclass
class Affected:
    """
    Class to hold affected data for a Vulnerability
    """

    remediation: Remediation
    package: Package = field(init=False)
    ranges: List[Range] = field(init=False) 

    def __post_init__(self):
        self.package = Package(self.remediation.component, self.remediation.cpe, self.remediation.purl)
        self.ranges = [Range(self.remediation.fixed_version)]
```

File: tools/redhat/osv.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        self.package = Package(remediation.component, remediation.cpe, remediation.purl)
+        self.ranges = [Range(remediation.fixed_version)]
+
+
+class OSV:
+    """
+    Class to convert CSAF data to OSV
+    """
+
+    SCHEMA_VERSION = "1.6.4"
```
Comment (andrewpollock) [inline tools/redhat/osv.py:104]: Update to currently released version

```suggestion
    SCHEMA_VERSION = "1.6.5"
```

File: tools/redhat/osv.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    ADVISORY_URL_PREFIXES = (
+        PKG_GO_DEV_VULN,
+        "https://www.cve.org/CVERecord",
+        "https://nvd.nist.gov/vuln/detail/",
+        "https://www.kb.cert.org/vuls/id/",
+        "https://github.com/advisories/",
+    )
+
+    def __init__(self, csaf_data: CSAF, modified: str, published: str = ""):
+        # Update this if verified against a later version
```
Comment (andrewpollock) [inline tools/redhat/osv.py:120]: This comment is probably better above line 104?

**Review by andrewpollock — CHANGES_REQUESTED** [review]
So I learned about this new dataclass hammer as part of reviewing this. Entirely your call as to whether you adjust to use it... The only definitely actionable items from me relate to schema version.

File: tools/redhat/osv.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+                        reference)
+        return [{"type": t, "url": u} for u, t in references.items()]
+
+    def _get_reference_type(self, reference: dict[str, str]) -> str:
+        """
+        Convert references from CSAF into typed referenced in OSV
+        Also make sure to add a related entry for any GO advisory references found
+        """
+        if reference["url"].startswith(ADVISORY_URL_PREFIXES):
+            self._add_go_related(reference["url"])
```
Comment (andrewpollock) [inline tools/redhat/osv.py:196]: This reads a bit awkwardly given it is called for non-Go advisories too.

```suggestion
            self._maybe_add_go_related(reference["url"])
```

File: tools/redhat/osv.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+            return "ADVISORY"
+        if reference["url"].startswith(
+                "https://bugzilla.redhat.com/show_bug.cgi"):
+            return "REPORT"
+        return "ARTICLE"
+
+    def _add_go_related(self, reference_url: str):
```
Comment (andrewpollock) [inline tools/redhat/osv.py:203]: Make it clearer that this is conditional behaviour

```suggestion
    def _maybe_add_go_related(self, reference_url: str):
```

File: tools/redhat/osv.py (hunk 9 above)
Comment (andrewpollock) [reply tools/redhat/osv.py:196]: Or maybe just pull out the conditional behaviour to the call site rather than inside the function?

Comment (jasinner) [reply tools/redhat/osv.py]: I removed this function and moved the code back in-line. I also renamed the calling function from `get_reference_type` to `get_reference_type_and_add_go_related` to better indicate it's dual behaviour.

**Review by andrewpollock — APPROVED** [review]

**Pull request merged**